File: sem3/graph/3a.py
n, m = map(int, input().split())
g = [set() for _ in range(n)]
for _ in range(m):
    a, b = map(int, input().split())
    g[a - 1].add(b - 1)
    g[b - 1].add(a - 1)
 
e = [999999] * n
r = [999999] * n
v = [False] * n
b = [set() for _ in range(n)]
def dfs(i, p=-1):
    v[i] = True
    e[i] = 0 if p == -1 else e[p] + 1
    r[i] = 0 if p == -1 else e[p] + 1
    for j in g[i]:
        if j == p:
            continue
        if v[j]:
            r[i] = min(r[i], e[j])
        else:
            dfs(j, i)
            r[i] = min(r[i], r[j])
            if r[j] > e[i]:
                b[i].add(j)
                b[j].add(i)
 
for i in range(n):
    if not v[i]:
        dfs(i)
 
c = [-1] * n
 
def dfs2(i, t):
    c[i] = t
    for j in b[i]:
        if c[j] == -1:
            dfs2(j, t)
 
t = 0
for i in range(n):
    if c[i] == -1:
        dfs2(i, t)
        t += 1
 
k = [-1] * t
for i in range(n):
    k[c[i]] += 1
 
r = 0
for i in range(n):
    r += k[c[i]] - len(b[i])
 
print(r // 2)

File: sem3/graph/3a.cpp
#include <bits/stdc++.h>

using namespace std;

vector<set<int>> g;
vector<set<int>> b;
vector<int> e;
vector<int> r;
vector<int> v;
vector<int> c;

void dfs(int i, int p = -1) {
    v[i] = true;
    e[i] = (p == -1) ? 0 : e[p] + 1;
    r[i] = (p == -1) ? 0 : e[p] + 1;
    for (auto j : g[i]) {
        if (j == p) continue;
        if (v[j]) {
            r[i] = min(r[i], e[j]);
        } else {
            dfs(j, i);
            r[i] = min(r[i], r[j]);
            if (r[j] > e[i]) {
                b[i].insert(j);
                b[j].insert(i);
            }
        }
    }
}

void dfs2(int i, int t) {
    c[i] = t;
    for (auto j : b[i]) {
        if (c[j] == -1) {
            dfs2(j, t);
        }
    }
}

int main() {
    int n, m;
    cin >> n >> m;
    g.resize(n);
    for (int i = 0; i < m; ++i) {
        int x, y;
        cin >> x >> y;
        g[x - 1].insert(y - 1);
        g[y - 1].insert(x - 1);
    }
    
    e.resize(n);
    r.resize(n);
    v.resize(n);
    b.resize(n);
    c.resize(n);
    for (int i = 0; i < n; ++i) {
        e[i] = 999999;
        r[i] = 999999;
        v[i] = false;
        c[i] = -1;
    }

    for (int i = 0; i < n; ++i) {
        if (!v[i]) {
            dfs(i);
        }
    }
    
    int t = 0;
    for (int i = 0; i < n; ++i) {
        if (c[i] == -1) {
            dfs2(i, t);
            t += 1;
        }
    }

    vector<int> k(t, -1);
    for (int i = 0; i < n; ++i) {
        k[c[i]] += 1;
    }

    long long res = 0;
    for (int i = 0; i < n; ++i) {
        res += k[c[i]] - b[i].size();
    }
    res /= 2;

    cout << res;
}
